William Bolder, Jongin Choi, Gianna Delorenzo, Naman Joshi, Ryan Kaye, and Darren Zhu  
DS 220   
Professor Fonkam  
December 7, 2023
# Rental Analysis of US Cities
## Introduction
This project aims to...  
First, we will import pandas and load the rental data

In [2]:
import pandas as pd

data = pd.read_csv('City_zori_uc_sfrcondomfr_sm_month.csv')

### Question 1: Which state has the highest average rent?  
Question 1 wants to know the state with the highest average rent and also the rent value for that state.  
1. First, identify the most recent date (`columns[-1]`) from the data set, which is October 2023.  
2. Second, initialize a dictionary to hold the total rent values for each state.  
3. Then, since we want to perform operations on each row of the data, we will use `iterrows()` in the for loop to iterate over each data frame row, which will yield an index and row data.  
4. The for loop will get the state and rent. If the rent is already in the dictionary `rent_sums`, it will add the rent value, else, it will initialize a new value.

In [19]:
most_recent = data.columns[-1]

# initialize dictionary
rent_sums = {}

# for each row in the data...
for index, row in data.iterrows():
    # get state and most recent rent
    state = row['State']
    rent = row[most_recent]

    # if the rent is already in dictionary add
    if state in rent_sums:
        rent_sums[state] += rent
        
    # else initialize
    else:
        rent_sums[state] = rent

5. Once the rent sum of each state is calculated, calculate the average rent for each state.
6. First, count the number of unique cities for each state with `value_counts()` and then initialize a dictionary to hold the average state rents.  
7. Then, in a for loop, for each state in the rent_sums dictionary, calculate the total rent, get the count of unique cities, calculate the average rent, and store the average rent in the `avg_state_rent` dictionary.
8. For the final step, the state (max key) and the corresponding rent value, then print out the results.


In [17]:
# count the number of unique "cities" for each state with value_counts()
state_rent_count = data['State'].value_counts()

# initialize dictionary
avg_state_rent = {}

# for each state in rent_sums...
for state in rent_sums:
    total_rent = rent_sums[state] # total rent for state
    city_count = state_rent_count[state] # get the count (number of cities) for state
    avg_rent = total_rent / city_count # calculate the average rent
    avg_state_rent[state] = avg_rent # store info in dictionary

# highest rent state and the average rent value
highest_rent_state = max(avg_state_rent, key = avg_state_rent.get)
highest_avg_rent = max(avg_state_rent.values())

print(f"The state with the highest average rent {highest_rent_state} with an average rent of ${highest_avg_rent:.2f}")

The state with the highest average rent NY with an average rent of $6153.50


#### Question 1 Results
The state with the highest average rent NY with an average rent of $6153.50

### Question 2: 